### PR TITLE
[Homo Optmization] Add homogeneous graph optimization 

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/construct_graph.py [hunk 1]
```diff
@@ -591,6 +591,16 @@ def verify_confs(confs):
             "The config file does not have a 'version' entry. Assuming gconstruct-v0.1")
     ntypes = {conf['node_type'] for conf in confs["nodes"]}
     etypes = [conf['relation'] for conf in confs["edges"]]
+    # Adjust input to DGL requirement if it is a honogeneous graph
+    if len(ntypes) == 1 and len(etypes) == 1 and not rev_edges:
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:595]: There can be multiple blocks under confs["nodes"] having the same ntype.
See https://github.com/awslabs/graphstorm/blob/main/tests/end2end-tests/data_process/data_gen.py#L155-L193 as an example.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 2]
```diff
@@ -582,7 +582,7 @@ def process_edge_data(process_confs, node_id_map, arr_merger,
 
     return (edges, edge_data, label_stats)
 
-def verify_confs(confs):
+def verify_confs(confs, rev_edges):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/construct_graph.py:585]: Please also add docstring for confs.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 3]
```diff
@@ -668,7 +685,7 @@ def process_graph(args):
             if args.num_processes_for_nodes is not None else args.num_processes
     num_processes_for_edges = args.num_processes_for_edges \
             if args.num_processes_for_edges is not None else args.num_processes
-    verify_confs(process_confs)
+    verify_confs(process_confs, args.add_reverse_edges)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:688]: How to handle the case if a user wants to add reverse edges?
We should have some thing like
```
etype = (th.cat((src, dst), th.cat(dst, src))
```
And also if there is edge features, we should at least crash the gconstruct.

File: tests/end2end-tests/data_process/movielens_test.sh [hunk 4]
```diff
@@ -27,6 +28,13 @@ python3 -m graphstorm.gconstruct.construct_graph --conf-file $GS_HOME/tests/end2
 
 error_and_exit $?
 
+echo "********* Test Homogeneous Graph Optimization ********"
```
Comment (classicsong) [inline tests/end2end-tests/data_process/movielens_test.sh:31]: Let's create a new .sh for homogeneous test, we can call it homogeneous_test.sh.

Comment (classicsong) [reply tests/end2end-tests/data_process/movielens_test.sh:31]: It is wired to put the test in movelens_test.sh.

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 5]
```diff
@@ -1705,6 +1705,58 @@ def test_gc():
     assert not os.path.isdir("/tmp_featurewrapper2"), \
         "Directory /tmp_featurewrapper2 should not exist after gc"
 
+
+def test_homo():
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1709]:
```suggestion
def test_homogeneous():
```

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 6]
```diff
@@ -1705,6 +1705,58 @@ def test_gc():
...
+    # single node type and edge type input
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1710]: You also need to test add reverse edges.

File: tests/end2end-tests/data_gen/process_movielens.py [hunk 7]
```diff
@@ -90,6 +90,11 @@ def write_data_parquet(data, data_file):
 edge_data = {'src_id': edges[0], 'dst_id': edges[1], 'rate': edges[2]}
 write_data_parquet(edge_data, '/data/ml-100k/edges.parquet')
 
+# generate data for homogeneous optimization test
+edges = pandas.read_csv('/data/ml-100k/u.data', delimiter='\t', header=None)
+edge_data = {'src_id': edges[1], 'dst_id': edges[1], 'rate': edges[2]}
+write_data_parquet(edge_data, '/data/ml-100k/edges_homo.parquet')
```
Comment (classicsong) [inline tests/end2end-tests/data_gen/process_movielens.py:96]: We can also generate the json config in process_movielens.py

File: tests/end2end-tests/data_gen/movielens_homo.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
+{
```
Comment (classicsong) [inline tests/end2end-tests/data_gen/movielens_homo.json:1]: Let's rename the file name to movie_homogeneous.json.

File: tests/end2end-tests/data_gen/process_movielens.py (hunk 7 above)
Comment (jalencato) [reply tests/end2end-tests/data_gen/process_movielens.py:96]: For our other gconstruct test cases, we directly use static json file stored in ```end2end-tests/data_gen```. We can keep the same way as what we did before.

File: tests/unit-tests/gconstruct/test_construct_graph.py (hunk 6 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_construct_graph.py:1710]: Refer to ```tests/end2end-tests/data_process/homogeneous_test.sh``` and ```tests/end2end-tests/data_process/check_homogeneous.py```

File: tests/end2end-tests/data_process/homogeneous_test.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+error_and_exit $?
+
+echo "********* Test Homogeneous Graph Optimization on reverse edge********"
+python3 -m graphstorm.gconstruct.construct_graph --conf-file $GS_HOME/tests/end2end-tests/data_gen/movielens_homogeneous.json --num-processes 1 --output-dir /tmp/movielen_100k_train_val_1p_4t_homogeneous_rev --graph-name movie-lens-100k --add-reverse-edges
+error_and_exit $?
+
+python3 $GS_HOME/tests/end2end-tests/data_process/check_homogeneous.py
+error_and_exit $?
+
+echo "********* Test Node Classification on GConstruct Homogeneous Graph on reverse edge********"
```
Comment (classicsong) [inline tests/end2end-tests/data_process/homogeneous_test.sh:44]:
```suggestion
echo "********* Test Node Classification on GConstruct Homogeneous Graph with reverse edge********"
```

File: tests/end2end-tests/data_process/homogeneous_test.sh [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+python3 -m graphstorm.run.gs_node_classification --workspace $GS_HOME/training_scripts/gsgnn_np/ --num-trainers $NUM_TRAINERS --num-servers 1 --num-samplers 0 --part-config /tmp/movielen_100k_train_val_1p_4t_homogeneous_rev/movie-lens-100k.json --ip-config ip_list.txt --ssh-port 2222 --cf ml_nc.yaml --target-ntype _N
+error_and_exit $?
+
+echo "********* Test Edge Classification on GConstruct Homogeneous Graph on reverse edge ********"
```
Comment (classicsong) [inline tests/end2end-tests/data_process/homogeneous_test.sh:48]:
```suggestion
echo "********* Test Edge Classification on GConstruct Homogeneous Graph with reverse edge ********"
```

File: tests/end2end-tests/data_process/movielens_test.sh [hunk 11]
```diff
@@ -5,8 +5,9 @@ service ssh restart
 GS_HOME=$(pwd)
 NUM_TRAINERS=4
 export PYTHONPATH=$GS_HOME/python/
+cd $GS_HOME/training_scripts/gsgnn_np
+echo "127.0.0.1" > ip_list.txt
```
Comment (classicsong) [inline tests/end2end-tests/data_process/movielens_test.sh:9]: Can we move these two to homogeneous_test.sh?

File: tests/unit-tests/gconstruct/test_construct_graph.py (hunk 6 above)
Comment (classicsong) [reply tests/unit-tests/gconstruct/test_construct_graph.py:1710]: Unit test can find the error earlier. If it is possible to write a unit test, you need to write a unit test.

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 12]
```diff
@@ -1705,6 +1705,46 @@ def test_gc():
     assert not os.path.isdir("/tmp_featurewrapper2"), \
         "Directory /tmp_featurewrapper2 should not exist after gc"
 
+
+def test_homogeneous():
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1709]: Please also add a test for is_homogeneous() function.

Comment (jalencato) [reply tests/unit-tests/gconstruct/test_construct_graph.py:1711]: Add the test here in the test_homogeneous() as we need to reuse the config here

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 13]
```diff
@@ -13,6 +13,7 @@
     See the License for the specific language governing permissions and
     limitations under the License.
 """
+import copy
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:16]: Two import copy

**Review by classicsong — APPROVED** [review]

**Pull request merged**